Deploy smart pointers in Position.cpp

rniwa opened the pull request:
#### 3660d70f9f4973c61dccfc581fd5ae6967a48546
<pre>
Deploy smart pointers in Position.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=263426">https://bugs.webkit.org/show_bug.cgi?id=263426</a>

Reviewed by Chris Dumez.

* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::rootEditableElementForPosition const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/dom/Position.cpp:
(WebCore::Position::containerOrParentElement const):
(WebCore::Position::anchorElementAncestor const):
(WebCore::Position::previous const):
(WebCore::Position::atFirstEditingPositionForNode const):
(WebCore::Position::atLastEditingPositionForNode const):
(WebCore::Position::parentEditingBoundary const):
(WebCore::Position::atStartOfTree const):
(WebCore::Position::atEndOfTree const):
(WebCore::Position::previousCharacterPosition const):
(WebCore::Position::nextCharacterPosition const):
(WebCore::isStreamer):
(WebCore::Position::upstream const):
(WebCore::Position::downstream const):
(WebCore::Position::nodeIsUserSelectAll):
(WebCore::Position::rootUserSelectAllForNode):
(WebCore::Position::isCandidate const):
(WebCore::Position::isRenderedCharacter const):
(WebCore::Position::rendersInDifferentPosition const):
(WebCore::Position::leadingWhitespacePosition const):
(WebCore::searchAheadForBetterMatch):
(WebCore::Position::inlineBoxAndOffset const):
(WebCore::commonInclusiveAncestor):
(WebCore::positionInParentBeforeNode):
(WebCore::positionInParentAfterNode):
(WebCore::Position::element const): Deleted.
* Source/WebCore/dom/Position.h:
(WebCore::Position::rootEditableElement const):
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::styleAtSelectionStart):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::findEventTargetFrom const):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::currentForm const):

Canonical link: <a href="https://commits.webkit.org/269571@main">https://commits.webkit.org/269571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b8fca15048816a699fec07edfa995de7ac78c98

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22924 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/951 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24022 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24835 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21223 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23190 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2118 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23460 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23165 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/462 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19880 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25692 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20759 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26963 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21018 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24817 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/459 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18261 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/375 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/871 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2903 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/626 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1599,15 +1599,14 @@ int AccessibilityRenderObject::indexForVisiblePosition(const VisiblePosition& po
-    Element* result = nullptr;
-    
-    Element* rootEditableElement = position.rootEditableElement();
+    RefPtr<Element> result = nullptr;
```
`= nullptr` can be omitted.

rniwa replied:
Removed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -207,7 +207,7 @@ RefPtr<Text> Position::protectedContainerText() const
 Element* Position::containerOrParentElement() const
 {
-    auto* container = containerNode();
+    Node* container = containerNode();
```
`auto*` was fine, no?

rniwa replied:
Reverted

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -337,12 +337,12 @@ Position::AnchorType Position::anchorTypeForLegacyEditingPosition(Node* anchorNo
     while (node && !is<Element>(*node))
         node = node->parentNode();
-    return downcast<Element>(node);
+    return downcast<Element>(node.get());
```
`downcast<Element>(WTFMove(node))` to avoid ref counting churn.

rniwa replied:
Fixed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -337,12 +337,12 @@ Position::AnchorType Position::anchorTypeForLegacyEditingPosition(Node* anchorNo
 
 // FIXME: This method is confusing (does it return anchorNode() or containerNode()?) and should be renamed or removed
-Element* Position::element() const
+RefPtr<Element> Position::anchorElementAncestor() const
```
This is fine but personally I hadn't been converting functions like these. It's all simple getters so it should be safe without smart pointers.

rniwa replied:
Alright, I'm gonna keep this change for now.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -477,8 +477,10 @@ bool Position::atFirstEditingPositionForNode() const
-        return !lastOffsetForEditing(*protectedDeprecatedNode());
+    case PositionIsAfterAnchor: {
+        auto node = protectedDeprecatedNode();
+        return !lastOffsetForEditing(*node);
```
I don't understand this change. It seems it was perfectly safe before? Why use a local variable if it is only used one time?

rniwa replied:
Reverted

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -489,7 +491,8 @@ bool Position::atLastEditingPositionForNode() const
     // FIXME: Position after anchor shouldn't be considered as at the first editing position for node since that position resides outside of the node.
-    return m_anchorType == PositionIsAfterAnchor || m_anchorType == PositionIsAfterChildren || m_offset >= static_cast<unsigned>(lastOffsetForEditing(*protectedDeprecatedNode()));
+    auto anchorNode = protectedDeprecatedNode();
+    return m_anchorType == PositionIsAfterAnchor || m_anchorType == PositionIsAfterChildren || m_offset >= static_cast<unsigned>(lastOffsetForEditing(*anchorNode));
```
Same comment, I believe the code was safe before and more concise. I personally prefer the old code.

rniwa replied:
Reverted.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -513,16 +516,16 @@ bool Position::atEditingBoundary() const
         return nullptr;
 
-    Node* boundary = m_anchorNode.get();
+    RefPtr boundary = m_anchorNode.get();
```
`.get()` can be omitted.

rniwa replied:
Removed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -548,8 +551,10 @@ bool Position::atStartOfTree() const
-        return !lastOffsetForEditing(*protectedAnchorNode());
+    case PositionIsAfterChildren: {
+        auto node = protectedAnchorNode();
+        return !lastOffsetForEditing(*node);
```
Same comment, I preferred the old code.

rniwa replied:
Reverted

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -571,8 +576,10 @@ bool Position::atEndOfTree() const
-        return !lastOffsetForEditing(*protectedAnchorNode());
+    case PositionIsBeforeChildren: {
+        auto node = protectedAnchorNode();
+        return !lastOffsetForEditing(*node);
```
Same comment, I preferred the old code.

rniwa replied:
Reverted

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -696,40 +704,41 @@ Position Position::upstream(EditingBoundaryCrossingRule rule) const
+    RefPtr boundary = enclosingVisualBoundary(startNode.get());
     // FIXME: PositionIterator should respect Before and After positions.
-    PositionIterator lastVisible = m_anchorType == PositionIsAfterAnchor ? makeDeprecatedLegacyPosition(m_anchorNode.get(), caretMaxOffset(*m_anchorNode)) : *this;
+    auto anchorNode = protectedAnchorNode();
```
Personally, I was using `RefPtr anchorNode = m_anchorNode;` or `RefPtr anchorNode = anchorNode();`. This is fine too but I felt the `protectedX()` functions were mostly useful to avoid local variables and pass the result directly as function parameter.

rniwa replied:
Okay, we can do: `RefPtr anchorNode = m_anchorNode;`

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -696,40 +704,41 @@ Position Position::upstream(EditingBoundaryCrossingRule rule) const
     bool boundaryCrossed = false;
     for (; !currentPosition.atStart(); currentPosition.decrement()) {
-        auto& currentNode = *currentPosition.node();
+        RefPtr currentNode = currentPosition.node();
```
Why aren't we using a `Ref` here to be closer to the old code and make it clear it can't be null?

rniwa replied:
Sure, changed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -802,50 +811,51 @@ Position Position::downstream(EditingBoundaryCrossingRule rule) const
+    RefPtr boundary = enclosingVisualBoundary(startNode.get());
     // FIXME: PositionIterator should respect Before and After positions.
-    PositionIterator lastVisible = m_anchorType == PositionIsAfterAnchor ? makeDeprecatedLegacyPosition(m_anchorNode.copyRef(), caretMaxOffset(*m_anchorNode)) : *this;
+    auto anchorNode = protectedAnchorNode();
```
Same comment as earlier about unnecessary call to a protected function.

rniwa replied:
Changed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -802,50 +811,51 @@ Position Position::downstream(EditingBoundaryCrossingRule rule) const
     bool boundaryCrossed = false;
     for (; !currentPosition.atEnd(); currentPosition.increment()) {
-        auto& currentNode = *currentPosition.node();
+        RefPtr currentNode = currentPosition.node();
```
Why not use Ref?

rniwa replied:
Changed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -969,24 +979,25 @@ bool Position::nodeIsUserSelectAll(const Node* node)
-        if (!nodeIsUserSelectAll(parent))
+        if (!nodeIsUserSelectAll(parent.get()))
             break;
         candidateRoot = parent;
```
`candidateRoot = WTFMove(parent);`

or

`candidateRoot = std::exchange(parent, candidateRoot->parentNode());` and omit the next line.

rniwa replied:
I'd go with WTFMove.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -1000,7 +1011,8 @@ bool Position::isCandidate() const
         return false;
 
-    auto* renderer = deprecatedNode()->renderer();
+    auto node = protectedDeprecatedNode();
```
Same comment as above about unnecessary call to protected function.

rniwa replied:
Changed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -1059,35 +1071,37 @@ bool Position::rendersInDifferentPosition(const Position& position) const
         return false;
 
-    auto* renderer = deprecatedNode()->renderer();
+    auto node = protectedDeprecatedNode();
```
same comment about unnecessary protected function call

rniwa replied:
Ditto.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -1059,35 +1071,37 @@ bool Position::rendersInDifferentPosition(const Position& position) const
-    
-    auto* positionRenderer = position.deprecatedNode()->renderer();
+
+    auto positionNode = position.protectedDeprecatedNode();
```
ditto.

rniwa replied:
Changed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -1143,8 +1153,10 @@ Position Position::leadingWhitespacePosition(Affinity affinity, bool considerNon
     Position prev = previousCharacterPosition(affinity);
-    if (prev != *this && inSameEnclosingBlockFlowElement(deprecatedNode(), prev.deprecatedNode()) && is<Text>(*prev.deprecatedNode())) {
-        UChar c = downcast<Text>(*prev.deprecatedNode()).data()[prev.deprecatedEditingOffset()];
+    auto node = protectedDeprecatedNode();
```
ditto.

cdumez replied:
With my proposal it is clear if we're dealing with a Ref or a RefPtr. It is also more concise.

rniwa replied:
OK. changed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -1143,8 +1153,10 @@ Position Position::leadingWhitespacePosition(Affinity affinity, bool considerNon
-    if (prev != *this && inSameEnclosingBlockFlowElement(deprecatedNode(), prev.deprecatedNode()) && is<Text>(*prev.deprecatedNode())) {
-        UChar c = downcast<Text>(*prev.deprecatedNode()).data()[prev.deprecatedEditingOffset()];
+    auto node = protectedDeprecatedNode();
+    auto previousNode = prev.protectedDeprecatedNode();
```
ditto.

rniwa replied:
Changed.

cdumez commented on Source/WebCore/dom/Position.cpp:
```diff
@@ -1278,7 +1290,8 @@ InlineBoxAndOffset Position::inlineBoxAndOffset(Affinity affinity, TextDirection
         box = textBox ? textBox : candidate;
     } else {
-        if (canHaveChildrenForEditing(*deprecatedNode()) && is<RenderBlockFlow>(*renderer) && hasRenderedNonAnonymousDescendantsWithHeight(downcast<RenderBlockFlow>(*renderer))) {
+        auto node = protectedDeprecatedNode();
```
ditto.

rniwa replied:
Changed.

cdumez reviewed (approved):
LGTM with comments.

rniwa committed:
Deploy smart pointers in Position.cpp
https://bugs.webkit.org/show_bug.cgi?id=263426

Reviewed by Chris Dumez.

* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::rootEditableElementForPosition const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/dom/Position.cpp:
(WebCore::Position::containerOrParentElement const):
(WebCore::Position::anchorElementAncestor const):
(WebCore::Position::previous const):
(WebCore::Position::atFirstEditingPositionForNode const):
(WebCore::Position::atLastEditingPositionForNode const):
(WebCore::Position::parentEditingBoundary const):
(WebCore::Position::atStartOfTree const):
(WebCore::Position::atEndOfTree const):
(WebCore::Position::previousCharacterPosition const):
(WebCore::Position::nextCharacterPosition const):
(WebCore::isStreamer):
(WebCore::Position::upstream const):
(WebCore::Position::downstream const):
(WebCore::Position::nodeIsUserSelectAll):
(WebCore::Position::rootUserSelectAllForNode):
(WebCore::Position::isCandidate const):
(WebCore::Position::isRenderedCharacter const):
(WebCore::Position::rendersInDifferentPosition const):
(WebCore::Position::leadingWhitespacePosition const):
(WebCore::searchAheadForBetterMatch):
(WebCore::Position::inlineBoxAndOffset const):
(WebCore::commonInclusiveAncestor):
(WebCore::positionInParentBeforeNode):
(WebCore::positionInParentAfterNode):
(WebCore::Position::element const): Deleted.
* Source/WebCore/dom/Position.h:
(WebCore::Position::rootEditableElement const):
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::styleAtSelectionStart):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::findEventTargetFrom const):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::currentForm const):

Canonical link: https://commits.webkit.org/269571@main
